### PR TITLE
Name video files more sensibly

### LINKED ISSUE #52: Name videos more sensibly
Never worked out how to change the output file name in OSN / OBS. 

Would be good to for a bunch of reasons.

### PR DESCRIPTION
**Pull request by tdaugaard**
This adds an easy and scalable way to find a raid name from an encounter ID; `raidInstances` is an array of raids with zone ID, name, and a list of encounters (by encounter ID).

Hashmap `instanceEncountersById` has been added to have a single point of truth regarding encounter IDs to encounter name for all instances that support this (dungeons, raids).

A new file has been added, `helpers.ts`, which should be kept FREE from any filesystem related modules (as these cannot be imported in the renderer process) and some of the helper functions from both `utils.ts` as well as `renderutils.ts` has been moved here, as they were needed in multiple places.

This file utilizes the format of `export const methodName = () => {};` as it avoids having a long list of exported names at the bottom, and makes it easy to identify which methods are exported and which aren't.

Furthermore, it vastly improves renaming symbols across the entire codebase because it isn't just changed to `export { newName as oldName }`.

The `Recorder` class will now generate a suitable filename for the final video file based on the given metadata to name the final video files a bit more sensibly.

The format that was chosen results in file names like:

  - `2022-09-22 15-36-33 - {2v2} Enigma Crucible (Win)`
  - `2022-09-22 16-15-54 - {Raids} Sepulcher of the First Ones, Lihuvim (Wipe)`
  - `2022-09-22 15-44-19 - {Mythic+} Mechagon Junkyard +18 (+1)`

So, basically: `{timestamp} - {category} {Zone/Instance}` + some category specific information as seen above, whenever possible.

Replaced all occurances of `{ [id: number]: string; }` with `NumberKeyToStringValueMapType` in `constants.ts`.

Closes #52

### CODE REVIEW

File: src/main/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import { dungeonsByMapId, instanceDifficulty, InstanceDifficultyType, instanceEncountersById, instanceNamesByZoneId, raidInstances, VideoCategory, zones } from "./constants";
+import { Metadata } from "./logutils";
+import { RaidInstanceType } from "./types";
+
```
Comment (aza547) [inline src/main/helpers.ts:4]: Can we get a comment in here to be explicit about no filesystem stuff in this file.

File: src/main/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * Get a result text appropriate for the video category that signifies a
+ * win or a loss, of some sort.
+ */
+export const getVideoResultText = (category: VideoCategory, isGoodResult: boolean): string => {
+
+    // Not sure how we can decide who won or lost yet.
```
Comment (aza547) [inline src/main/helpers.ts:11]: This probably needs an update. "Non-trivial to determine who won a BG/SoloShuffle so just don't report it" -- something like that

File: src/main/recorder.ts [hunk 3]
```diff
@@ -280,6 +289,52 @@ type RecorderOptionsType = {
...
+     * to identify the video files on the filesystem and be able to tell what it
+     * was about and the result of it.
+     */
+    getFinalVideoFilename (metadata: Metadata): string | undefined {
+        let outputFilename = '';
+        let zoneName: string;
+        const resultText = getVideoResultText(metadata.category, metadata.result);
+
+        switch (metadata.category) {
+            case VideoCategory.TwoVTwo:
```
Comment (aza547) [inline src/main/recorder.ts:307]: Can we use default to avoid this pattern of multiple cases in a row

File: src/main/recorder.ts [hunk 4]
```diff
@@ -280,6 +289,52 @@ type RecorderOptionsType = {
...
+                    outputFilename = `${getDungeonByMapId(cm.mapId)} +${cm.level} (${resultText})`;
+                }
+            break;
+        }
+
+        if (!outputFilename) {
+            return;
+        }
+
+        return `{${metadata.category}} ${outputFilename}`;
```
Comment (aza547) [inline src/main/recorder.ts:336]: Not a big fan of curly braces making it to something user visible. Think it would be better to be:

```
2022-09-22 20-33-47 - 2v2 - Enigma Crucible (Win)
2022-09-22 20-30-39 - Mythic+ - Mechagon Junkyard +18 (+1)
```
Spaces in file names is weird to me as well but I think that's just business as usual on windows so I'll get over it!

File: src/main/util.ts [hunk 5]
```diff
@@ -453,10 +421,16 @@ const runSizeMonitor = async (storageDir: string, maxStorageGB: number): Promise
...
+const cutVideo = async (
+        initialFile: string,
+        finalDir: string,
+        outputFilename: string | undefined,
+        desiredDuration: number
+    ): Promise<string> => {
     
     const videoFileName = path.basename(initialFile, '.mp4');
-    const finalVideoPath = path.join(finalDir, videoFileName + ".mp4");
+    const baseVideoFilename = sanitizeFilename(videoFileName + (outputFilename ? ' - ' + outputFilename : ''));
```
Comment (aza547) [inline src/main/util.ts:432]: Doing this inline is a bit hard to read. Can we bring it out?

```
if (outputFilename) {
  outputFilename = " - " + outputFilename;
} else {
  outputFilename = "";
}

const baseVideoFilename = sanitizeFilename(videoFileName + outputFilename);
```

File: src/main/recorder.ts [hunk 6]
```diff
@@ -280,6 +289,52 @@ type RecorderOptionsType = {
...
+        let zoneName: string;
+        const resultText = getVideoResultText(metadata.category, metadata.result);
+
+        switch (metadata.category) {
+            case VideoCategory.TwoVTwo:
+            case VideoCategory.ThreeVThree:
+            case VideoCategory.Skirmish:
+            case VideoCategory.SoloShuffle:
+                zoneName = getInstanceNameByZoneId(metadata.zoneID);
+                outputFilename = `${zoneName} (${resultText})`;
```
Comment (aza547) [inline src/main/recorder.ts:312]: No resultText for SoloShuffles
Also BGs, but they are missed out of this function, I think by mistake?

File: src/main/recorder.ts [hunk 7]
```diff
@@ -280,6 +289,52 @@ type RecorderOptionsType = {
...
+            case VideoCategory.Raids:
+                const encounterName = getEncounterNameById(metadata.encounterID);
+                zoneName = getRaidNameByEncounterId(metadata.encounterID);
+                outputFilename = `${zoneName}, ${encounterName} (${resultText})`;
+            break;
+
+            case VideoCategory.MythicPlus:
+                if (metadata.challengeMode) {
+                    const cm = metadata.challengeMode;
+                    const keystoneUpgradeLevel = ChallengeModeDungeon.calculateKeystoneUpgradeLevel(cm.allottedTime, cm.duration);;
```
Comment (aza547) [inline src/main/recorder.ts:324]: This case could probably do with its own helper function `getChallengeModeResultText` or so, there is a bit much packed in here for my liking. VideoButton could also benefit from sharing it?

File: src/main/util.ts [hunk 8]
```diff
@@ -442,6 +398,18 @@ const runSizeMonitor = async (storageDir: string, maxStorageGB: number): Promise
     writeMetadataFile(videoPath, metadata);
 }
 
+/**
```
Comment (aza547) [inline src/main/util.ts:401]: Love this!

**Review by aza547 — CHANGES_REQUESTED** [review]
Few minor comments, looks great.

File: src/main/recorder.ts (hunk 4 above)
Comment (tdaugaard) [reply src/main/recorder.ts:336]: Spaces in filenames is totally normal on Windows, yeah :D Curly braces look bad, agreed.

File: src/main/recorder.ts (hunk 3 above)
Comment (tdaugaard) [reply src/main/recorder.ts:307]: We can - I know I did that somewhere else, but for this one I just felt it made more sense to have the explicitly stated to make it quickly obvious what was being handled how, but ultimately I don't really care either way =)

File: src/main/recorder.ts (hunk 4 above)
Comment (tdaugaard) [reply src/main/recorder.ts:336]: Spaces in file names are here to stay :p Curly braces looks bad, I agree.

File: src/main/recorder.ts (hunk 7 above)
Comment (tdaugaard) [reply src/main/recorder.ts:324]: VideoButton wouldn't benefit from this, as it's only used for the filename, really. It's not so much result text (which we have already "_Depl_" and "_Time_") as it is a condensed string of information for the given video.

I can surely give it its own method!

File: src/main/recorder.ts (hunk 6 above)
Comment (tdaugaard) [reply src/main/recorder.ts]: Most likely because I know nothing about PvP. I made a change now where, if `resultText` is empty it won't add it to the filename - will that satisfy SoloShuffles and BGs?

Comment (aza547) [reply src/main/recorder.ts]: Sounds good. The problem with solo shuffle is that there are 6 round and you can win any number of that. BGs there is just no indication in logs who wins. Currently in either case we don't display a result.
